config/config.go

```diff
@@ -90,56 +90,83 @@ func parsePlugins(pluginList []interface{}) ([]*PluginConfig, error) {
 	return plugins, nil
 }
 
-func (c *Config) parseV6Config() error {
-	if exists := c.v.Get("server6"); exists == nil {
-		// it is valid to have no DHCPv6 configuration defined, so no
-		// server and no error are returned
-		return nil
+func (c *Config) getListenAddress(v6 bool) (*net.UDPAddr, error) {
+	ver := 6
+	if !v6 {
+		ver = 4
+	}
+	if exists := c.v.Get(fmt.Sprintf("server%d", ver)); exists == nil {
+		// it is valid to have no server configuration defined, and in this case
+		// no listening address and no error are returned.
+		return nil, nil
 	}
-	addr := c.v.GetString("server6.listen")
+	addr := c.v.GetString(fmt.Sprintf("server%d.listen", ver))
 	if addr == "" {
-		return ConfigErrorFromString("dhcpv6: missing `server6.listen` directive")
+		return nil, ConfigErrorFromString("dhcpv%v: missing `server%d.listen` directive", ver, ver)
 	}
 	ipStr, portStr, err := net.SplitHostPort(addr)
 	if err != nil {
-		return ConfigErrorFromString("dhcpv6: %v", err)
+		return nil, ConfigErrorFromString("dhcpv%d: %v", ver, err)
 	}
 	ip := net.ParseIP(ipStr)
-	if ip.To4() != nil {
-		return ConfigErrorFromString("dhcpv6: missing or invalid `listen` address")
+	if v6 && ip.To4() != nil {
+		return nil, ConfigErrorFromString("dhcpv%d: missing or invalid `listen` address", ver)
+	} else if !v6 && ip.To4() == nil {
+		return nil, ConfigErrorFromString("dhcpv%d: missing or invalid `listen` address", ver)
 	}
 	port, err := strconv.Atoi(portStr)
 	if err != nil {
-		return ConfigErrorFromString("dhcpv6: invalid `listen` port")
+		return nil, ConfigErrorFromString("dhcpv%d: invalid `listen` port", ver)
 	}
 	listener := net.UDPAddr{
 		IP:   ip,
 		Port: port,
 	}
-	sc := ServerConfig{
-		Listener: &listener,
-		Plugins:  nil,
-	}
-	// load plugins
+	return &listener, nil
+}
+
+func (c *Config) getPlugins(v6 bool) ([]*PluginConfig, error) {
 	pluginList := cast.ToSlice(c.v.Get("server6.plugins"))
 	if pluginList == nil {
-		return ConfigErrorFromString("dhcpv6: invalid plugins section, not a list")
+		return nil, ConfigErrorFromString("dhcpv6: invalid plugins section, not a list")
```

Comment:
The comment mentions dhcpv6 which may not always be the case.

Reply:
I'm not familiar with the code-base, but I also find it curious that `server6.plugins` is used here, and in the original code there is no `server4.plugins` counterpart.  Is this something that is shared between v4/v6, but is for some reason stored under server6 by default?

Reply:
it's a copypasta error, it should be protocol-specific, not "dhcpv6". This PR is changing probably too many things at the same time, because it's restructuring the config parsing code to be reusable, but also adding the v4 parsing.